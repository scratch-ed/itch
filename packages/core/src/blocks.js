/* Copyright (C) 2019 Ghent University - All Rights Reserved */
import { isNumber } from 'lodash-es/lang.js';

export function containsLoop(blocks) {
  for (const key in blocks) {
    if (key === 'control_repeat' || key === 'control_forever') return true;
  }
  return false;
}

export function containsBlock(name, blocks) {
  for (const key in blocks) {
    if (key === name) return true;
  }
  return false;
}

export function countExecutions(name, blocks) {
  for (const key in blocks) {
    if (key === name) return blocks[key];
  }
  return 0;
}

function getOrNull(blockId, blockmap, sprite) {
  if (blockId) {
    const parentBlock = blockmap.get(blockId);
    return blockToNode(parentBlock, blockmap, sprite);
  } else {
    return null;
  }
}

function convertInput(inputArray, blockmap, sprite) {
  // We ignore shadows, as they are not that relevant for us.
  // As such, we always convert the second element in the input array.
  if (Array.isArray(inputArray[1])) {
    const input = inputArray[1];
    return input[1];
  } else {
    // ID of a block.
    const id = inputArray[1];
    return blockToNode(id, blockmap, sprite);
  }
}

function convertMutation(mutation) {
  if (!mutation) {
    return null;
  }

  return mutation.proccode;
}

/**
 * Convert one block to a tree node.
 *
 * @param {Sb3Block} block
 * @param {Sb3Target} sprite
 * @param {Map<string,Sb3Block>} blockmap
 *
 * @return {Object}
 */
function blockToNode(block, blockmap, sprite) {
  const next = getOrNull(block.next, blockmap, sprite);

  const input = {};
  for (const [key, value] of Object.entries(block.inputs || {})) {
    input[key] = convertInput(value, blockmap, sprite);
    if (isNumber(input[key])) {
      input[key] = input[key].toString();
    }
  }

  return {
    opcode: block.opcode,
    next: next,
    input: input,
    mutation: convertMutation(block.mutation),
  };
}

/**
 * Convert all blocks from a sprite to a list of trees, where each tree is
 * a set of attached blocks.
 *
 * @param {Sb3Target} sprite
 * @param {Sb3Block[]} blocks - Top level blocks to filter.
 *
 * @return {Set<Object>}
 */
export function asTree(
  sprite,
  blocks = sprite.blocks.filter((b) => b.topLevel),
) {
  const blockMap = new Map(sprite.blocks.map((i) => [i.id, i]));

  // Find all top-level blocks.
  const filtered = blocks
    .filter((b) => b.topLevel)
    .map((b) => blockToNode(b, blockMap, sprite));

  return new Set(filtered);
}
