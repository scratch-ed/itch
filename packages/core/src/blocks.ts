/* Copyright (C) 2019 Ghent University - All Rights Reserved */
import isNumber from 'lodash-es/isNumber';
import { Sb3Block, Sb3Mutation, Sb3Target } from './structures';

/**
 * @deprecated
 */
export function containsLoop(blocks: Record<string, any>): boolean {
  for (const key in blocks) {
    if (key === 'control_repeat' || key === 'control_forever') return true;
  }
  return false;
}

/**
 * @deprecated
 */
export function containsBlock(name: string, blocks: object): boolean {
  for (const key in blocks) {
    if (key === name) return true;
  }
  return false;
}

/**
 * @deprecated
 */
export function countExecutions(name: string, blocks: Record<string, number>): number {
  for (const key in blocks) {
    if (key === name) return blocks[key];
  }
  return 0;
}

function getOrNull(blockId: string | null, blockmap: Map<string, Sb3Block>): Node | null {
  if (blockId) {
    const parentBlock = blockmap.get(blockId)!;
    return blockToNode(parentBlock, blockmap);
  } else {
    return null;
  }
}

function convertInput(inputArray: any[], blockmap: Map<string, Sb3Block>) {
  // We ignore shadows, as they are not that relevant for us.
  // As such, we always convert the second element in the input array.
  if (Array.isArray(inputArray[1])) {
    const input = inputArray[1];
    return input[1];
  } else {
    // ID of a block.
    const id = inputArray[1];
    return blockToNode(id, blockmap);
  }
}

function convertMutation(mutation: Sb3Mutation | null): string | null {
  if (!mutation) {
    return null;
  }

  return mutation.proccode;
}

interface Node {
  opcode: string;
  next: Node | null;
  input: Record<string, any>;
  mutation: string | null;
}

/**
 * Convert one block to a tree node.
 */
function blockToNode(block: Sb3Block, blockmap: Map<string, Sb3Block>): Node {
  const next = getOrNull(block.next, blockmap);

  const input: Record<string, any> = {};
  for (const [key, value] of Object.entries(block.inputs || {})) {
    input[key] = convertInput(value, blockmap);
    if (isNumber(input[key])) {
      input[key] = input[key].toString();
    }
  }

  return {
    opcode: block.opcode,
    next: next,
    input: input,
    mutation: convertMutation(block.mutation),
  };
}

/**
 * Convert all blocks from a sprite to a list of trees, where each tree is
 * a set of attached blocks.
 *
 * @param sprite
 * @param blocks - Top level blocks to filter.
 */
export function asTree(
  sprite: Sb3Target,
  blocks: Sb3Block[] = sprite.blocks.filter((b) => b.topLevel)
): Set<object> {
  const blockMap = new Map(sprite.blocks.map((i) => [i.id, i]));

  // Find all top-level blocks.
  const filtered = blocks
    .filter((b) => b.topLevel)
    .map((b) => blockToNode(b, blockMap));

  return new Set(filtered);
}
