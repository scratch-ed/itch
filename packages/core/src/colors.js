/* Copyright (C) 2019 Ghent University - All Rights Reserved */
export function detectColor(logData) {
    return logData.color;
}