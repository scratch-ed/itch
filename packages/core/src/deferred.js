/**
 * A "Deferred" like implementation on top of a Promise.
 */
export default class Deferred {
  constructor() {
    /**
     * The promise. Use this to await completion.
     * @type {Promise<any>}
     */
    this.promise = new Promise((resolve, reject) => {
      /**
       * Call to resolve the underlying promise.
       * @type {function(any): void}
       */
      this.resolve = resolve;
      /**
       * Call to reject the underlying promise.
       * @type {function(Error): void}
       */
      this.reject = reject;
    });
  }
}