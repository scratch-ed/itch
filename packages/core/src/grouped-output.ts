/**
 * @fileOverview A way to group tests in a flexible hierarchy. In the long term,
 *   this will replace the existing groups.
 */

/**
 * Handle outputting. By default, all output is sent to stderr.
 * You can overwrite this by setting a global `handleOut` callback,
 * which will be used instead if present.
 */
function toOutput(output: Record<string, unknown>): void {
  // @ts-ignore
  if (typeof window.handleOut !== 'undefined') {
    // @ts-ignore
    window.handleOut(output);
  } else {
    console.log(output);
  }
}

export enum Status {
  TimeLimit = 'time limit exceeded',
  Runtime = 'runtime error',
  Wrong = 'wrong',
  Correct = 'correct',
}

export enum Format {
  Text = 'text',
  Html = 'html',
}

export interface OutputHandler {
  (obj: Record<string, unknown>): void;
}

export interface Message {
  format: Format;
  description: string;
}

export enum Visibility {
  Show = 'show',
  Collapse = 'collapse',
}

/**
 * Manages the output for the Dodona-inspired format.
 *
 * While this class is exposed in testplans, in most cases
 * you should use the high-level testplan API instead of this one.
 *
 * ### Dodona format
 *
 * Some more information on the Dodona format. The format is a partial format.
 * The judge basically sends updates to the test result state via commands, e.g.
 * "start testcase X", "start test Y", "close testcase X", etc.
 *
 * For ease of use, the result manager will automatically open higher levels when
 * opening lower levels. For example, if you open a testcase without opening a
 * context first, the result manager will do so for you. Previous levels are also
 * closed when appropriate. For example, when starting a new tab, all previous tabs
 * will be closed.
 *
 * There is one exception: a test. If an open test is detected, an error will be thrown,
 * as the result manager has no way of knowing if the test is successful or not.
 */
export class GroupedResultManager {
  private readonly out: OutputHandler;
  private hasOpenJudgement = false;
  private openGroups = 0;
  private hasOpenTest = false;
  private isFinished = false;
  private escalation?: Status;

  constructor(out: OutputHandler = toOutput) {
    this.out = out;
  }

  /**
   * Start the judgement.
   */
  startJudgement(): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to start judgement after judgement has been completed. Ignoring.',
      );
      return;
    }
    this.out({ command: 'start-judgement', version: 2 });
    this.hasOpenJudgement = true;
  }

  /**
   * Close the judgement. This will finish the judge, meaning all future
   * output is ignored.
   */
  closeJudgement(): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to close judgement after judgement has been completed. Ignoring.',
      );
      return;
    }
    if (!this.hasOpenJudgement) {
      console.warn('Attempting to close judgement while none is open. Ignoring.');
      return;
    }
    while (this.openGroups > 0) {
      console.warn('Unclosed groups, force closing them...');
      this.closeGroup();
    }
    this.out({ command: 'close-judgement' });
    this.hasOpenJudgement = false;
    this.isFinished = true;
  }

  /**
   * Start a new test group.
   *
   * Each group needs a name, which is user visible. You can also optionally
   * provide the name of a sprite, which allows for a better display, for
   * example by showing the sprite or linking to it. You must use the sprite's
   * name.
   *
   * @param name Name of the test group.
   * @param [sprite] Name of optionally linked sprite.
   */
  startGroup(name: string, sprite?: string): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to open group after judgement has been completed. Ignoring.',
      );
      return;
    }
    if (!this.hasOpenJudgement) {
      this.startJudgement();
    }
    this.openGroups++;
    this.out({ command: 'start-group', name: name, sprite: sprite });
  }

  /**
   * Close a group.
   *
   * @param summary
   * @param visibility
   */
  closeGroup(summary?: string | Message, visibility: Visibility = Visibility.Show): void {
    if (this.isFinished) {
      console.warn(
        'Attempting to close group after judgement has been completed. Ignoring.',
      );
      return;
    }
    if (this.openGroups <= 0) {
      console.warn('Attempting to close tab while none is open. Ignoring.');
      return;
    }

    this.openGroups--;
    this.out({ command: 'close-group', summary: summary, visibility: visibility });
  }

  startTest(name = `Test ${Math.random()}`): void {
    if (this.isFinished) {
      console.warn('Attempting to start test after judgement has been completed.');
      return;
    }
    if (this.hasOpenTest) {
      console.warn(
        'Attempting to start new test while open test exists, closing it as wrong.',
      );
      this.closeTest(Status.Wrong);
    }
    if (this.openGroups <= 0) {
      console.warn('Attempting to start test while no group is open.');
      this.startGroup(`Groep ${Math.random()}`);
    }
    this.out({
      command: 'start-test',
      name: name,
    });
    this.hasOpenTest = true;
  }

  closeTest(status: Status, description?: string): void {
    if (this.isFinished) {
      console.warn('Attempting to close test after judgement has been completed.');
      return;
    }
    if (!this.hasOpenTest) {
      console.warn('Attempting to close test while none is open. Ignoring.');
      return;
    }
    this.out({
      command: 'close-test',
      status: status,
      description: description,
    });
    this.hasOpenTest = false;
  }

  appendMessage(message: string): void {
    if (!this.hasOpenJudgement) {
      console.warn('Attempting to append message while no judgement is open. Ignoring.');
      return;
    }
    this.out({ command: 'append-message', message: message });
  }

  appendDiff(expected: string, actual: string | null = null): void {
    if (!this.hasOpenTest) {
      console.warn('Attempting to append diff while no test is open. Ignoring.');
      return;
    }
    this.out({ command: 'append-diff', expected: expected, actual: actual });
  }

  escalateStatus(status: Status | string): void {
    if (!this.hasOpenJudgement) {
      console.warn('Attempting to escalate status of closed judgement. Ignoring.');
      return;
    }
    this.escalation = status as Status;
    this.out({ command: 'escalate-status', status: status });
  }

  whyEscalate(): Status | undefined {
    return this.escalation;
  }
}
