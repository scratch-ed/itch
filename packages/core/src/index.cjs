// Wrapper for common js modules.
const result = require('./evaluation.js');

module.exports = result.run;
