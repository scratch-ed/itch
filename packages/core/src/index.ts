// Wrapper for common js modules.
import { run } from './evaluation';
export default run;
