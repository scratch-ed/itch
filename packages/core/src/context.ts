import type Runtime from 'itch-scratch-vm-types/types/engine/runtime';
import type Thread from 'itch-scratch-vm-types/types/engine/thread';
import type VirtualMachine from 'scratch-vm';

import { Deferred } from './deferred';
import { BroadcastReceiver, ThreadListener } from './listener';
import { Event, Log } from './log';
import { OutputHandler, ResultManager } from './output';
import { ProfileEventData } from './profiler';
import { proxiedRenderer, RendererMethods, unproxyRenderer } from './renderer';
import { EndAction } from './scheduler/end';
import { ScheduledEvent } from './scheduler/scheduled-event';
import { memoize } from './utils';

// A list of event names...
// These must be the same as in Scratch.
const SCRATCH_PROJECT_START = 'PROJECT_START';
const SCRATCH_PROJECT_RUN_STOP = 'PROJECT_RUN_STOP';
const SCRATCH_SAY_OR_THINK = 'SAY';
const SCRATCH_QUESTION = 'QUESTION';
const SCRATCH_ANSWER = 'ANSWER';
// Custom events we emit ourselves.
const STOPPED_THREAD = 'STOPPED_THREAD';
const BEFORE_HATS_START = 'BEFORE_HATS_START';
// Custom event for the debugger.
export const OPS = 'OPS_EXECUTED';

/**
 * @typedef {object} Acceleration
 * @property {number} factor - Main acceleration factor.
 * @property {number} [time] - Scratch waiting times.
 * @property {number} [event] - Scheduled events waiting time.
 */
interface Acceleration {
  factor: number;
  time?: number;
  event?: number;
}

interface VmMethods {
  step: Runtime['_step'];
  startHats: Runtime['startHats'];
  stopThreads: Runtime['_stopThread'];
}

interface EventHandles {
  scratchProjectStart: () => void;
  // eslint-disable-next-line
  scratchSayOrThink: (...args: any[]) => void;
  // eslint-disable-next-line
  scratchQuestion: (...args: any[]) => void;
  // eslint-disable-next-line
  scratchAnswer: (...args: any[]) => void;
  // eslint-disable-next-line
  scratchProjectRunStop: (...args: any[]) => void;
  threadStopped: (thread: Thread) => void;
  // eslint-disable-next-line
  beforeHatsStart: (...args: any[]) => void;
  // eslint-disable-next-line
  ops: (...args: any[]) => void;
}

type LogMode = 'judge' | 'debugger';

/**
 * Contains common information and parameters for the
 * judge. This is passed around in lieu of using globals.
 */
export class Context {
  readonly vm: VirtualMachine;
  numberOfRun: number;
  readonly newLog: Log;
  answers: string[];
  providedAnswers: string[];
  /**
   * Resolves once the simulation has ended.
   */
  simulationEnd: Deferred<string>;
  /**
   * The timeout for the actions.
   */
  actionTimeout: number;
  /**
   * The listeners for the threads.
   */
  threadListeners: ThreadListener[];
  broadcastListeners: BroadcastReceiver[];
  event: ScheduledEvent;
  groupedOutput: ResultManager;

  /**
   * The acceleration factor, used to speed up (or slow down)
   * execution in the VM.
   *
   * There is a limit on how much you can increase this, since
   * each step in the VM must still have time to run of course.
   */
  accelerationFactor: Acceleration;

  vmMethods?: VmMethods;
  eventHandles?: EventHandles;
  renderMethods?: RendererMethods;
  // eslint-disable-next-line @typescript-eslint/ban-types
  blockMethods?: Record<string, Function>;

  constructor(vm: VirtualMachine, log: Log, callback?: OutputHandler) {
    this.numberOfRun = 0;
    this.answers = [];
    this.providedAnswers = [];
    this.simulationEnd = new Deferred();
    this.actionTimeout = 5000;
    this.threadListeners = [];
    this.broadcastListeners = [];
    this.event = ScheduledEvent.create();
    this.groupedOutput = new ResultManager(callback);
    this.accelerationFactor = {
      factor: 1,
    };
    this.vm = vm;
    this.newLog = log;
  }

  /**
   * Get a current timestamp.
   * @return {number}
   * @deprecated Use the log instead.
   */
  timestamp(): number {
    return this.log.timestamp();
  }

  private interceptVmMethods(): void {
    if (this.vmMethods) {
      throw new Error('VM methods already intercepted');
    }

    const oldStepFunction = this.vm.runtime._step.bind(this.vm.runtime);
    this.vm.runtime._step = () => {
      const oldResult = oldStepFunction();
      this.vm.runtime._lastStepDoneThreads.forEach((t) =>
        this.vm.runtime.emit(STOPPED_THREAD, t),
      );
      return oldResult;
    };

    const oldThreadFunction = this.vm.runtime.sequencer.retireThread.bind(
      this.vm.runtime.sequencer,
    );
    this.vm.runtime.sequencer.retireThread = (thread: Thread): void => {
      this.vm.emit(STOPPED_THREAD, thread);
      oldThreadFunction(thread);
    };

    const oldHatFunction = this.vm.runtime.startHats.bind(this.vm.runtime);
    this.vm.runtime.startHats = (requestedHatOpcode, optMatchFields, optTarget) => {
      this.vm.runtime.emit(BEFORE_HATS_START, {
        requestedHatOpcode,
        optMatchFields,
        optTarget,
      });
      return oldHatFunction(requestedHatOpcode, optMatchFields, optTarget);
    };

    this.vmMethods = {
      step: oldStepFunction,
      startHats: oldHatFunction,
      stopThreads: oldThreadFunction,
    };
  }

  private restoreVmMethods(): void {
    if (!this.vmMethods) {
      console.log('VM methods not intercepted, doing nothing.');
      return;
    }
    this.vm.runtime._stopThread = this.vmMethods.stopThreads;
    this.vm.runtime.startHats = this.vmMethods.startHats;
    this.vm.runtime._step = this.vmMethods.step;
    this.vmMethods = undefined;
  }

  /**
   * Attach event handles to the VM.
   *
   * @param logMode - The mode in which logging should happen. When using
   *   the "judge" mode, we listen to various input events, but we don't listen
   *   to the custom "ops" event, which is only emitted by our forked VM.
   */
  private attachEventHandles(logMode: LogMode = 'judge'): void {
    if (this.eventHandles) {
      throw new Error('Event handles already attached');
    }

    this.eventHandles = {
      scratchProjectStart: () => {
        console.log(`${this.log.timestamp()}: run number: ${this.numberOfRun}`);
        this.numberOfRun++;
      },
      scratchSayOrThink: (target, type, text) => {
        // Only save it when something is actually being said.
        if (text !== '') {
          console.log(`${this.log.timestamp()}: say: ${text} with ${type}`);
          const event = new Event('say', {
            text: text,
            target: target,
            type: type,
            sprite: target.sprite.name,
          });
          event.previous = this.log.snap('event.say');
          event.next = event.previous;
          this.log.registerEvent(event);
        }
      },
      scratchQuestion: (question) => {
        if (question != null) {
          let x = this.providedAnswers.shift();
          if (x === undefined) {
            this.groupedOutput.appendMessage(
                'Er werd een vraag gesteld waarop geen antwoord voorzien is.',
            );
            this.groupedOutput.escalateStatus('wrong');
            x = undefined;
          }

          console.log(`${this.log.timestamp()}: input: ${x}`);

          const event = new Event('answer', {
            question: question,
            text: x,
          });
          event.previous = this.log.snap('event.answer');
          event.next = event.previous;
          this.log.registerEvent(event);

          this.vm.runtime.emit(SCRATCH_ANSWER, x);
        }
      },
      // The following is only used by the debugger
      scratchAnswer: (answer) => {
        const event = new Event('answer', {
          text: answer,
        });
        event.previous = this.log.snap('event.answer');
        event.next = event.previous;
        this.log.registerEvent(event);
      },
      scratchProjectRunStop: () => {
        for (const action of this.threadListeners) {
          if (action.active) {
            action.onStopped();
          }
        }
        console.log(`${this.log.timestamp()}: Ended run`);
      },
      threadStopped: (thread) => {
        for (const action of this.threadListeners) {
          if (action.active) {
            action.onThreadStopped(thread);
          }
        }
      },
      beforeHatsStart: (opts) => {
        if (opts.requestedHatOpcode === 'event_whenbroadcastreceived') {
          this.broadcastListeners
            .filter((l) => l.active)
            .forEach((l) =>
              l.update({
                matchFields: opts.optMatchFields,
                target: opts.optTarget,
              }),
            );
        }
      },
      ops: (ops) => {
        if (ops.length === 0) {
          return;
        }
        const block = ops[ops.length - 1]; // Only log the last outer block

        // @ts-ignore
        const isMonitored = this.vm.runtime.monitorBlocks.getBlock(block.id)?.isMonitored;
        if (isMonitored) {
          return;
        }

        const event = new Event('ops', {
          blockId: block.id,
        });
        event.previous = this.log.snap('event.ops');
        event.next = event.previous;
        this.log.registerEvent(event);
      },
    };

    if (logMode === 'judge') {
      console.log('Attaching event listeners for judge log mode...');
      this.vm.runtime.on(SCRATCH_PROJECT_START, this.eventHandles.scratchProjectStart);
      this.vm.runtime.on(SCRATCH_SAY_OR_THINK, this.eventHandles.scratchSayOrThink);
      this.vm.runtime.on(SCRATCH_QUESTION, this.eventHandles.scratchQuestion);
      this.vm.runtime.on(
        SCRATCH_PROJECT_RUN_STOP,
        this.eventHandles.scratchProjectRunStop,
      );
      this.vm.runtime.on(STOPPED_THREAD, this.eventHandles.threadStopped);
      this.vm.runtime.on(BEFORE_HATS_START, this.eventHandles.beforeHatsStart);
    }

    if (logMode === 'debugger') {
      console.log('Attaching event listeners for debugger log mode...');
      this.vm.runtime.on(OPS, this.eventHandles.ops);
      this.vm.runtime.on(SCRATCH_ANSWER, this.eventHandles.scratchAnswer);
    }
  }

  public setLogRange(start: number, end: number): void {
    this.log.setRange(start, end);
  }

  /**
   * Detach event handles from the VM.
   */
  private detachEventHandles(): void {
    if (!this.eventHandles) {
      console.log('No event handles found to detach, doing nothing.');
      return;
    }
    this.vm.runtime.off(SCRATCH_PROJECT_START, this.eventHandles.scratchProjectStart);
    this.vm.runtime.off(SCRATCH_SAY_OR_THINK, this.eventHandles.scratchSayOrThink);
    this.vm.runtime.off(SCRATCH_QUESTION, this.eventHandles.scratchQuestion);
    this.vm.runtime.off(
      SCRATCH_PROJECT_RUN_STOP,
      this.eventHandles.scratchProjectRunStop,
    );
    this.vm.runtime.off(STOPPED_THREAD, this.eventHandles.threadStopped);
    this.vm.runtime.off(BEFORE_HATS_START, this.eventHandles.beforeHatsStart);
    this.vm.runtime.off(OPS, this.eventHandles.ops);
    this.eventHandles = undefined;
  }

  /**
   * Create a profile and attach it to the VM.
   */
  private createProfiler(): void {
    this.vm.runtime.enableProfiling();
    const blockId = this.vm.runtime.profiler.idByName('blockFunction');
    this.vm.runtime.profiler.onFrame = (frame) => {
      if (frame.id === blockId) {
        this.log.snap('profiler.basic');
      }
    };
  }

  /** @internal */
  public createAdvancedProfiler(onlyThese?: string[]) {
    if (this.blockMethods) {
      throw new Error('Profiler already created');
    }

    console.log('Installing advanced block profiler...');
    // Attach the advanced profiler.
    this.blockMethods = {};
    // Allow use inside the nested function below.
    const log = this.log;
    for (const [opcode, blockFunction] of Object.entries(this.vm.runtime._primitives)) {
      this.blockMethods[opcode] = blockFunction;
      this.vm.runtime._primitives[opcode] = new Proxy(blockFunction, {
        apply: function (target, thisArg, argumentsList) {
          const vmTarget = argumentsList[1].target;
          const targetName = vmTarget.getName();
          // Only track sprites we care about if given.
          if (onlyThese === undefined || onlyThese.includes(targetName)) {
            const currentBlockId = argumentsList[1].thread.peekStack();
            // Only register block executions that exist; other blocks we don't care about.
            if (vmTarget.blocks.getBlock(currentBlockId)) {
              const data: ProfileEventData = {
                blockId: currentBlockId,
                target: targetName,
                block: memoize(() => {
                  const target = log.last.target(targetName);
                  return target.block(currentBlockId);
                }),
                node: memoize(() => {
                  const target = log.last.target(targetName);
                  return target.node(currentBlockId);
                }),
              };
              const event = new Event('block_execution', data);
              event.previous = log.last;
              event.next = event.previous;
              log.registerEvent(event);
            }
          }

          return target.apply(thisArg, argumentsList);
        },
      });
    }
  }

  /** @internal */
  public removeAdvancedProfiler() {
    if (!this.blockMethods) {
      console.log('No profiler found to remove, doing nothing.');
      return;
    }

    for (const [opcode, blockFunction] of Object.entries(this.blockMethods)) {
      this.vm.runtime._primitives[opcode] = blockFunction;
    }
  }

  private removeProfiler(): void {
    this.vm.runtime.disableProfiling();
  }

  private proxyRenderer(): void {
    if (this.renderMethods) {
      throw new Error('Renderer already proxied');
    }

    this.renderMethods = proxiedRenderer(this.log, this.vm.renderer);
  }

  private unproxyRenderer(): void {
    if (!this.renderMethods) {
      console.log('No renderer found to unproxy, doing nothing.');
      return;
    }
    unproxyRenderer(this.vm.renderer, this.renderMethods);
    this.renderMethods = undefined;
  }

  /**
   * Instrument the VM, based on the log mode.
   *
   * @param logMode
   */
  public instrumentVm(logMode: LogMode = 'judge'): void {
    this.interceptVmMethods();
    this.attachEventHandles(logMode);
    this.proxyRenderer();
    if (logMode === 'judge') {
      this.createProfiler();
    }
  }

  /**
   * Remove all instrumentations from the VM.
   * This should restore the VM to its original state.
   */
  public deinstrumentVm(): void {
    this.removeProfiler();
    this.unproxyRenderer();
    this.detachEventHandles();
    this.restoreVmMethods();
    this.removeAdvancedProfiler();
  }

  /**
   * Prepare the VM for execution for the judge.
   *
   * This will prepare the answers for questions (if applicable) and instrument
   * the VM to take the acceleration factor into account.
   */
  public runVm(): void {
    this.providedAnswers = this.answers.slice();

    // Optimisation.
    if (this.accelerationFactor.time !== 1) {
      // We need to instrument the VM.
      // Changing the events is not necessary; this
      // is handled by the event scheduler itself.

      // First, modify the step time.
      const currentStepInterval = (<typeof Runtime>this.vm!.runtime.constructor)
        .THREAD_STEP_INTERVAL;
      const newStepInterval = currentStepInterval / this.accelerationFactor.factor;

      Object.defineProperty(this.vm!.runtime.constructor, 'THREAD_STEP_INTERVAL', {
        value: newStepInterval,
      });

      // We also need to change various time stuff.
      this.acceleratePrimitive('control_wait', 'DURATION');
      this.acceleratePrimitive('looks_sayforsecs');
      this.acceleratePrimitive('looks_thinkforsecs');
      this.acceleratePrimitive('motion_glidesecstoxy');
      this.acceleratePrimitive('motion_glideto');

      this.accelerateTimer();
    }

    // Start the vm.
    this.vm.start();
  }

  /**
   * Adjust the given argument for a given opcode to the acceleration factor.
   *
   * This is used to modify "time" constants to account for the acceleration factor.
   * For example, if a condition is "wait for 10 seconds", but the acceleration factor
   * is 2, we only want to wait for 5 seconds, not 10.
   *
   * @param opcode - The opcode to accelerate.
   * @param argument - The argument to accelerate.
   *
   * @private
   */
  acceleratePrimitive(opcode: string, argument = 'SECS'): void {
    const original = this.vm.runtime.getOpcodeFunction(opcode);
    const factor = this.accelerationFactor.time || this.accelerationFactor.factor;
    this.vm.runtime._primitives[opcode] = (
      originalArgs: Record<string, unknown>,
      util: unknown,
    ) => {
      // For safety, clone the arguments.
      const args = { ...originalArgs };
      args[argument] = <number>args[argument] / factor;
      return original(args, util);
    };
  }

  /**
   * Adjust the given method on the given device to account for the
   * acceleration factor.
   *
   * This is mainly used to reverse accelerate the project timer.
   * E.g. if the project timer is counts 10s for a project with
   * acceleration factor 2, it should count 20s instead.
   */
  accelerateTimer(): void {
    const factor = this.accelerationFactor.time || this.accelerationFactor.factor;
    const device = this.vm.runtime.ioDevices.clock;
    const original = device.projectTimer;

    device.projectTimer = () => {
      return original.call(device) * factor;
    };
  }

  /**
   * Accelerate a certain number. This is intended for events.
   *
   * @param number - The number to accelerate. All non-numbers are returned as is.
   */
  accelerateEvent<T>(number: number | T): number | T {
    const factor = this.accelerationFactor.event || this.accelerationFactor.factor;
    if (factor === 1 || typeof number !== 'number') {
      return number;
    }
    return number / factor;
  }

  terminate(): void {
    const action = new EndAction();
    action.execute(this, () => {});
  }

  get log(): Log {
    return this.newLog!;
  }
}
