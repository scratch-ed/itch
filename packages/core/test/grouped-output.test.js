import { GroupedResultManager, Status } from '../src/grouped-output';

test('normal usage', () => {
  const results = [];
  const manager = new GroupedResultManager((r) => results.push(r));

  manager.startJudgement();
  manager.startGroup('Groep 1', 'test-sprite');
  manager.startTest('Hallo one test...');
  manager.appendMessage('Hallo');
  manager.appendDiff('expected', 'actual');
  manager.closeTest(Status.Wrong, 'Wrong');
  manager.closeGroup('A nice summary');
  manager.closeJudgement();

  expect(results).toMatchSnapshot();
});
