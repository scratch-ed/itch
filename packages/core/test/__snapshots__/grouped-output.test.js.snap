// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normal usage 1`] = `
Array [
  Object {
    "command": "start-judgement",
    "version": 2,
  },
  Object {
    "command": "start-group",
    "name": "Groep 1",
    "sprite": "test-sprite",
    "visibility": "show",
  },
  Object {
    "command": "start-test",
    "name": "Hallo one test...",
  },
  Object {
    "command": "append-message",
    "message": "Hallo",
  },
  Object {
    "actual": "actual",
    "command": "append-diff",
    "expected": "expected",
  },
  Object {
    "command": "close-test",
    "description": "Wrong",
    "status": "wrong",
  },
  Object {
    "command": "close-group",
    "summary": "A nice summary",
  },
  Object {
    "command": "close-judgement",
  },
]
`;
